### ISSUE TITLE
Catastrophic errors in new build on ORNL Summit, last commit built 4 months ago still builds and runs fine

### ISSUE THREAD

**Issue by asedova** [cp2k/cp2k#1771]
Hi all- I am rebuilding CP2K on Summit after our OS upgrade. In August, I had a well running build using git:678222b. Now during the build, I error out with `Error: Unused variable 'p_work_tmp' declared at (1) [-Werror=unused-variable]` but can build if I remove that flag from the arch files. However, even with a totally serial, no GPU test I get rapid NaNs and infinity for all regtests I have tried. When I check out that previous commit, everything is fine and the tests pass.

Details:
GCC: I  tried 9, 10, and 7. I used 7 for the one that worked with `git:678222b`
CUDA: 11.0.3
FFTW: I used both system and local build through the toolchain. The final build that works used system
Openblas: Tried 0.3.5 (this one is built with no threading) from system, local build with toolchain, final working build used openblas/0.3.17-omp (with threads) which is the default system version
I used no other libraries, built with no packages except DBCSR
Scalapack and lapack are the netlib refs

It seems to me that it's not an issue with the libraries on Summit, based on the fact that `678222b` is fine.

arch file is this (with the `-Werror=unused-variable` flag removed from `WFLAGS` for the current version)
```
CC          = mpicc
CXX         = mpicxx
AR          = ar -r
FC          = mpif90
LD          = mpif90
#
DFLAGS      = -D__parallel -D__DBCSR_ACC -D__PW_CUDA  -D__FFTW3  -D__SCALAPACK -D__ACC 
#
WFLAGS      = -Werror=aliasing -Werror=ampersand -Werror=c-binding-type -Werror=intrinsic-shadow -Werror=intrinsics-std -Werror=line-truncation -Werror=tabs -Werror=target-lifetime -Werror=underflow -Werror=unused-but-set-variable -Werror=unused-variable -Werror=unused-dummy-argument -Werror=conversion -Werror=zerotrip -Wno-maybe-uninitialized -Wuninitialized -Wuse-without-only 
#
FCDEBFLAGS  = -fbacktrace -ffree-form -fimplicit-none -std=f2008 
CFLAGS      = -mtune=native -fno-omit-frame-pointer -O3 -funroll-loops $(PROFOPT) -fopenmp  -std=c11 -Wall -Wextra -Werror $(DFLAGS)
FCFLAGS     =  -O2 -funroll-loops  -fopenmp  $(FCDEBFLAGS) $(WFLAGS) $(DFLAGS)
#
LDFLAGS     =  $(FCFLAGS) -Wl,--enable-new-dtags -L${OLCF_FFTW_ROOT}/lib -L${OLCF_NETLIB_LAPACK_ROOT}/lib -L${OLCF_NETLIB_SCALAPACK_ROOT}/lib -L${OLCF_CUDA_ROOT}/lib64 -L${OLCF_OPENBLAS_ROOT}/lib -L'/usr/lib64' -Wl,-rpath='/usr/lib64'
LIBS     = -L${OLCF_FFTW_ROOT}/lib -lfftw3 -lfftw3_threads \
           -L${OLCF_NETLIB_LAPACK_ROOT}/lib  -llapack \
           -L${OLCF_NETLIB_SCALAPACK_ROOT}/lib -lscalapack \
           -L${OLCF_CUDA_ROOT}/lib64 -L${OLCF_CUDA_ROOT}/include -L${OLCF_CUDA_ROOT}/bin -lcudart -lcuda -lstdc++ -lnvrtc \
           -L${OLCF_OPENBLAS_ROOT}/lib -lblas -lfftw3_mpi -lfftw3 -lfftw3_omp  -lopenblas -lstdc++  -lcufft -lcublas -lrt 
#
GPUVER        = V100
OFFLOAD_CC    = nvcc
OFFLOAD_FLAGS = -g -arch sm_70 -O3 -Xcompiler='-fopenmp' --std=c++11 $(DFLAGS) 
#
FYPPFLAGS   = -n --line-marker-format=gfortran5
```

**Comment by abussy** [comment]
It looks like PR #1798 should fix this issue (included in the latest release, i.e. CP2K 9.1)

**Comment by asedova** [comment]
Just got a chance to try the new version. NaNs continue:
```
 McWeeny: Deviation of idempotency      8.09653177
  McWeeny: Deviation of idempotency    896.24527598
  McWeeny: Deviation of idempotency****************
 Energy with the initial guess:********************

  McWeeny: Deviation of idempotency****************
  McWeeny: Deviation of idempotency****************
  McWeeny: Deviation of idempotency        Infinity
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
 
  BCH: step  1 Norm of P_old-Pnew:      0.00000000
 
  McWeeny: Deviation of idempotency      0.00000000

 SCF     1     -3803.527983313********************    3.874944

  BCH: step  1 Norm of P_old-Pnew:      0.00000000
  McWeeny: Deviation of idempotency      0.00000000

 SCF     2     -3803.527983313********************    2.980913

  Line Search: Step Size      0.45000000 Predicted energy********************
  BCH: step  1 Norm of P_old-Pnew:      0.00000000
  McWeeny: Deviation of idempotency      0.00000000

 SCF     3     -3803.527983313********************    2.979087

  Convergence of Newton iteration    46.964906
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
  Convergence of Newton iteration          NaN
 
  BCH: step  1 Norm of P_old-Pnew:      0.00000000
 
```

arch file:
```
CC          = mpicc
CXX         = mpicxx
AR          = ar -r
FC          = mpif90
LD          = mpif90
#

DFLAGS      = -D__GRID_CUDA -D__parallel -D__DBCSR_ACC  -D__FFTW3  -D__SCALAPACK -D__ACC
#
WFLAGS      = -Werror=aliasing -Werror=ampersand -Werror=c-binding-type -Werror=intrinsic-shadow -Werror=intrinsics-std -Werror=line-truncation -Werror=tabs -Werror=target-lifetime -Werror=underflow -Werror=unused-but-set-variable  -Werror=unused-variable -Werror=unused-dummy-argument -Werror=conversion -Werror=zerotrip -Wno-maybe-uninitialized -Wuninitialized -Wuse-without-only
#
FCDEBFLAGS  = -fbacktrace -ffree-form -fimplicit-none -std=f2008
CFLAGS      = -mtune=native -fno-omit-frame-pointer -O3 -funroll-loops $(PROFOPT) -fopenmp  -std=c11 -Wall -Wextra -Werror $(DFLAGS)
FCFLAGS     =  -O2 -funroll-loops  -fopenmp  $(FCDEBFLAGS) $(WFLAGS) $(DFLAGS)
#
LDFLAGS     =  $(FCFLAGS) -Wl,--enable-new-dtags -L${OLCF_FFTW_ROOT}/lib -L${OLCF_ESSL_ROOT}/lib64/ -lessl  -L${OLCF_NETLIB_LAPACK_ROOT}/lib  -llapack -lblas -L${OLCF_NETLIB_SCALAPACK_ROOT}/lib -L${OLCF_CUDA_ROOT}/lib64  -L'/usr/lib64' -Wl,-rpath='/usr/lib64'

LIBS     = -L${OLCF_FFTW_ROOT}/lib -lfftw3 -lfftw3_threads -L${OLCF_ESSL_ROOT}/lib64/ -L${OLCF_NETLIB_LAPACK_ROOT}/lib  -llapack -L${OLCF_NETLIB_SCALAPACK_ROOT}/lib -lscalapack -L${OLCF_CUDA_ROOT}/lib64 -L${OLCF_CUDA_ROOT}/include -L${OLCF_CUDA_ROOT}/bin -lcudart -lcuda -lstdc++ -lnvrtc -lfftw3_mpi -lfftw3 -lfftw3_omp  -lstdc++  -lcufft -lcublas -lrt
#
GPUVER        = V100
OFFLOAD_CC    = nvcc

OFFLOAD_TARGET = cuda

OFFLOAD_FLAGS = -arch sm_70 -O3 -Xcompiler='-fopenmp' --std=c++11 $(DFLAGS)
#
FYPPFLAGS   = -n --line-marker-format=gfortran5

```

modules:
essl/6.3.0 gcc/9.1.0 cuda/11.0.3 fftw/3.3.9 netlib-lapack/3.9.1 netlib-scalapack/2.1.0 python/3.8-anaconda3

**Comment by asedova** [comment]
Same behavior with reference netlib only for all blas.

**Comment by asedova** [comment]
Ok, so it seems this doesn't happen with gcc 7.5 on our system. But something about gcc 9 is causing it. So far all tests I have run have passed with the gcc 7.5/essl/cuda build. Gromacs and Amber seem to run fine on our systems with gcc 9 and cuda  10 or 11 and the same fftw, so I'm thinking it's not our libraries/compiler.

**Comment by asedova** [comment]
You can see in N2_t01.inp where the SCF values don't seem to move at all:

```
    28 P_Mix/Diag. 0.40E+00    0.0     1.13531904       -20.9600112749  5.70E-03
    29 P_Mix/Diag. 0.40E+00    0.0     1.13530965       -20.9656900663 -5.68E-03
    30 P_Mix/Diag. 0.40E+00    0.0     1.13531490       -20.9600011334  5.69E-03
    31 P_Mix/Diag. 0.40E+00    0.0     1.13531167       -20.9656840350 -5.68E-03
    32 P_Mix/Diag. 0.40E+00    0.0     1.13531348       -20.9599975508  5.69E-03
    33 P_Mix/Diag. 0.40E+00    0.0     1.13531236       -20.9656819040 -5.68E-03
    34 P_Mix/Diag. 0.40E+00    0.0     1.13531299       -20.9599962848  5.69E-03
    35 P_Mix/Diag. 0.40E+00    0.0     1.13531260       -20.9656811508 -5.68E-03
    36 P_Mix/Diag. 0.40E+00    0.0     1.13531282       -20.9599958373  5.69E-03
    37 P_Mix/Diag. 0.40E+00    0.0     1.13531269       -20.9656808845 -5.69E-03
    38 P_Mix/Diag. 0.40E+00    0.0     1.13531276       -20.9599956790  5.69E-03
    39 P_Mix/Diag. 0.40E+00    0.0     1.13531271       -20.9656807903 -5.69E-03
    40 P_Mix/Diag. 0.40E+00    0.0     1.13531274       -20.9599956231  5.69E-03
    41 P_Mix/Diag. 0.40E+00    0.0     1.13531272       -20.9656807570 -5.69E-03
    42 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9599956033  5.69E-03
    43 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9656807453 -5.69E-03
    44 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9599955963  5.69E-03
    45 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9656807411 -5.69E-03
    46 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9599955938  5.69E-03
    47 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9656807396 -5.69E-03
    48 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9599955929  5.69E-03
    49 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9656807391 -5.69E-03
    50 P_Mix/Diag. 0.40E+00    0.0     1.13531273       -20.9599955926  5.69E-03

```
then this happens:

```

  Iteration    Method   Restart      Stepsize      Convergence         Time
  ------------------------------------------------------------------------------
        1        PCG       F         0.00E+00    162.8122114151        0.87
        2        PCG       F         0.10E+01  ****************        0.91
        3        PCG       F         0.10E+01  ****************        0.94
        4        PCG       F         0.10E+01  ****************        0.98
        5        PCG       F         0.10E+01  ****************        1.01
        6        PCG       F         0.10E+01  ****************        1.04
        7        PCG       F         0.10E+01  ****************        1.08
        8        PCG       F              NaN               NaN        1.12
        9        PCG       F              NaN               NaN        1.13
       10        PCG       F              NaN               NaN        1.15
       11        PCG       F              NaN               NaN        1.16
       12        PCG       F              NaN               NaN        1.18
       13        PCG       F              NaN               NaN        1.20
       14        PCG       F              NaN               NaN        1.21
       15        PCG       F              NaN               NaN        1.23
       16        PCG       F              NaN               NaN        1.24
       17        PCG       F              NaN               NaN        1.26
       18        PCG       F              NaN               NaN        1.28
```

I let this run for 200 SCF steps and it stays locked on one value.

**Comment by oschuett** [comment]
@asedova, is this issue still present or can we close it?

**Comment by asedova** [comment]
@oschuett it seems to still be an issue, but I need to rebuild with the latest version to double check. However, I am wondering if this is related to a very similar issue happening on our AMD GPU test systems and on NERSC Perlmutter (NVIDIA A100s)-- same input test (H2O-curvy-prop.inp). I'm getting segfaults when using about 8 threads with the GPU version of DBCSR on Perlmutter. I just tried it last week but with v 9.1.0 release. On Spock, I was using @mtaillefumier 's build after having a lot of trouble with segfaults with DBCSR-GPU, and I also got segfaults with his build, but now I realize I was testing with 8 threads there too. Last week I realized that the segfaults stop on Perlmutter when I use 1 thread, and it made me realize this may be something that goes beyond the new AMD GPU versions.
